File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/Visitor/Visitor.lua
-- 访问者接口
local IComputerPartVisitor = {}

function IComputerPartVisitor:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

function IComputerPartVisitor:visitComputer(computer)
    error("Subclass must implement visitComputer method")
end

function IComputerPartVisitor:visitMouse(mouse)
    error("Subclass must implement visitMouse method")
end

function IComputerPartVisitor:visitKeyboard(keyboard)
    error("Subclass must implement visitKeyboard method")
end

function IComputerPartVisitor:visitMonitor(monitor)
    error("Subclass must implement visitMonitor method")
end

return IComputerPartVisitor

File: DesignPatterns/01_GOF23/DP_examples/Python/03_Behavioral/Visitor/Visitor.py
# 访问者接口
from abc import ABC, abstractmethod

class IComputerPartVisitor(ABC):
    @abstractmethod
    def visit_computer(self, computer):
        pass
    
    @abstractmethod
    def visit_mouse(self, mouse):
        pass
    
    @abstractmethod
    def visit_keyboard(self, keyboard):
        pass
    
    @abstractmethod
    def visit_monitor(self, monitor):
        pass
